Grouping servers as tree view - completed

wacki4 opened the pull request:
Completed saving and viewing for tree folders view.

TODO:
- i don't know how to correct edit icon for folders in server configuration view,
- some strange 'box' inside folder name - i don't know where does it get there
- overall check for such big change ;)

Regarding to #72

cschindl commented:
Hi @wacki4, 

thanks for your help. I'll look into it in the next few days. It would be good if you could fix the merge conflicts.

wacki4 commented:
I think is done. I made it by page, so could be some code mistake, but i think that everything is allright.

wacki4 commented on lib/helper/storage.js:
```diff
@@ -0,0 +1,270 @@
+
+             tree_folders.forEach((item, index) => {
+                 if (item.parent !== null) {
+                     tree_folders[map[item.parent]].childrens.push(item);
```
Could not reproduce it, made it safe.

wacki4 commented on lib/helper/storage.js:
```diff
@@ -0,0 +1,270 @@
+        return self.servers;
+    }
+
+    getTree() {
```
I use it in tree view - i need to build somewhere a tree, so i thought that te best place to make it would be Storage helper, than tree view. Maybe we would need it in future somewhere else.

wacki4 commented on lib/views/folder-view.js:
```diff
@@ -0,0 +1,156 @@
+    self.isExpanded = false;
+
+    self.name = self.config.name;
+    self.childrens = self.config.childrens;
```
I do not understand - everywhere i used childrens, not children, it is array of childrens.

wacki4 replied:
Allright, i've checked dictionary ;)

wacki4 commented on lib/views/folder-view.js:
```diff
@@ -0,0 +1,156 @@
+
+      // Events
+      server.getConnector().on('log', (msg) => {
+        self.ftpLogView.addLine(msg);
```
I don't really know what to do :) So i removed ftplogger, and if You think that something need to change - be free for change.

wacki4 commented:
So i put some corrections, resolved  checks i've set to resolved, not - made a comment. I think that i done everything.

Additional to Your message:

Enhancement:
* Check source code formatting (.jsbeautifyrc) - **i think that done, is there any way to 'beautify entire project' in atom?**
* Treeview: sort folders before servers - **done, easy, i thought that we will allow to sort for user as they want (by name), but maybe it is better option, so i only removed resorting on end,**
* Folder structure: check circular relations (actually folder A can be a parent of folder B, but at the same time folder B can also be parent of folder A) - **as i think i've done fully circular check. It won't allow user to make such error,**

Error:
* Disable drag/drop handler for folders - **Done**
* Changes on global storage object are used in tree view even if i doesn't save the configuration. - **TODO - any idea how to make it now?**
* Folder options in select will not be loaded if i open the configuration view. After i click the edit button and go back, the options will be added. - **TODO - as i understand You mean reloading folders after edit them? Or something else?**
* After saving the configuration the folder and servers will not be added. - **TODO - always? in file or on the view?**
* When I click the edit folder button in the configuration and go back, the first server in the list is selected. This is confusing :) - **Done**

wacki4 commented:
I have idea for 
* Changes on global storage object are used in tree view even if i doesn't save the configuration.

Maybe if clicked cancle/close window reload from config all data? And on save - always save it? It would be simplest way to do it.

cschindl commented:
Your idea could work like this. I will check and test your changes promptly.

cschindl commented:
Most of the adjustments look good. Some things can still be optimized but I could do that later. During testing I noticed some mistakes:

folder configuration
- If no folders have been created (which is the case when I open the configuration for the first time), the input fields in the configuration folder should not be editable either. The fields should only be editable after clicking on new and generating a new entry.
- After I create a folder for the first time and click on save and return to the server configuration, the folder I just created is not offered as a selection. This is only available after a restart.
- Circular relations between the folders still exist. E.g. If you click on new 2 times, 2 folders will be created. Now I switch to folder 1 and assign folder 2 as parents. Now I switch to folder 2 and assign folder 1 as parents, which should not work. When saving it comes to an error because there is a circle relation.
- When I delete a folder, the first folder in the list is selected. Unfortunately, the data in the input fields are not updated. They still show the data of the deleted folder.
- Your idea for changes on global storage object would not work. For example, open the server configuration and change something on the currently selected server. Now click on Edit at the folder button, create a new folder and click on save. Now the folder has been saved to the configuration. Unfortunately also the changes I made to the server before. If I now click on cancel the changes are still saved on the server what is wrong. The only idea I have is to work with a copy of the storage object in the configuration and only save it when saving the configuration. The storage object should reload the new configuration.

treeview:
-  Keyboard shortcut (right key) doesn't open the file or expand the folder like in the master version. 
   Line 683 Wrong:  if (!current.view().isExpanded) {
   Line 683 OK:  if (!current.view().isExpanded()) {
 
ftp-remote-editable.js
- found self.info.password. Please check. I think this is included by merging again.

wacki4 commented:
* Your idea for changes on global storage object would not work. For example, open the server configuration and change something on the currently selected server. Now click on Edit at the folder button, create a new folder and click on save. Now the folder has been saved to the configuration. Unfortunately also the changes I made to the server before. If I now click on cancel the changes are still saved on the server what is wrong. The only idea I have is to work with a copy of the storage object in the configuration and only save it when saving the configuration. The storage object should reload the new configuration.

About that - maybe simpler would be saving only folder part when folder editing? Because now, in Your way - it looks for me like if i save Folder, and after that cancle server config - it won't be save as i would predict. Maybe make a copy of servers for configuration view, and also copy of folders for folder configuration view? But there it could became a bug when we delete folder on folders view... This case look really hard as i see it. Maybe, we shouldn't combine this in current way, but make 'edit folders' on context menu on remote pane?

cschindl commented:
And if we implement my suggestion and remove in the folder configuration the save butten. Then you could only manage the folders there but saving the whole configuration, including the server settings, would only happen in the server configuration. This would make everything logical again.

wacki4 commented:
So we only have 'Close button' on Folder Configuriation view, and if we made some error, we close this window and close ConfigurationView, and reopen it, and everything is allright after that? It looks good and logical.

Only hard for now would be copy of Storage object, but it probably only look so hard ;)

wacki4 commented:
So maybe if we don't have save button over Folder Configuration View, reloading after closing and not saving ConfigurationView would be ok? It will make everything easier.

cschindl commented:
Yes it would be ok. The result should be the same.

wacki4 commented:
So i will try to complete it this week. Some of problems i've resolved localy.

wacki4 commented:
I am to stubborn to get to sleep with such a problems in my head ;) I think that now everything.

I will only remember You problem with button next to folder chooser, and strange box inside folder name input ;) I don't know how where is there a problem.

cschindl commented on lib/views/configuration-view.js:
```diff
@@ -893,6 +906,7 @@ export default class ConfigurationView extends View {
     }
 
     folderConfigurationView.on('close', (e) => {
+      self.createControlsFolderSelect();
```
Unfortunately, reloading the options causes the previous selection from the current selected server will be lost.

cschindl commented:
Circular relations between the folders still exist. I Think best place would be in folder-configuration-view.js in fillInputFields function.

wacki4 commented:
Circular relations between the folders still exist

Can You tell me how to reproduce circular relations? I've tried yesterday many combination but nothing allowed mi such a change.

wacki4 commented on lib/views/configuration-view.js:
```diff
@@ -520,10 +531,10 @@ export default class ConfigurationView extends View {
-      }
+    self.createControlsFolderSelect();
+
+    while (self.selectServer.firstChild) {
```
I've read around internet yesterday and this was always best way to do it. Are You shure about Your way?

wacki4 replied:
I will leave it this way - if You would like to change it, feel free. For me this code is shorter and cleaner, i'm not shure if faster but maybe.

cschindl commented:
E.g. If you click on new 2 times, 2 folders will be created. Now I switch to folder 1 and assign folder 2 as parents. Now I switch to folder 2 and assign folder 1 as parents, which should not work. When saving it comes to an error because there is a circle relation.

wacki4 commented:
Ok i will try at home to reproduce and correct it.

wacki4 commented on lib/ftp-remote-edit.js:
```diff
@@ -438,7 +438,7 @@ class FtpRemoteEdit {
     const self = this;
 
-    if (!self.info.password) {
+    if (Storage.hasPassword()) {
```
I think that You checked one commit before - it has been changed.

wacki4 commented on lib/ftp-remote-edit.js:
```diff
@@ -428,7 +428,7 @@ class FtpRemoteEdit {
     const self = this;
 
-    if (!self.info.password) {
+    if (Storage.hasPassword()) {
```
I think that You checked one commit before - it has been changed.

cschindl replied:
Wrong: if (Storage.hasPassword()) {
OK: if (!Storage.hasPassword()) {

wacki4 replied:
Now i understand, done ;)

wacki4 commented:
I didn't close review requesting which i didn't change/didn't found on code. Please check on Your side.

I think that now we are really close.

cschindl commented:
Thanks for your help. Everything looks good so far. I will try to solve the connection problems from the issues first. After that I will make the last adjustments. You helped me a lot. Thank you very much for that.

wacki4 commented:
Always ready to help, cannot wait for this release ;) Next things coming from me :)

wacki4 commented:
When do You plan release ? :)

cschindl commented:
Unfortunately, it took a little longer to fix the connection errors. But now that they seem to have been fixed, I can take care of the implementation of new features. So I will try to start this week. I don't have an exact date for the release yet. It will be released when it is finished) ;)

wacki4 commented:
I know, i also make everything in my free time ;) I have now only problem cause on my home computer i have new config, so everytime i have to run atom from console :D And also waiting for all changes on master, to make new ones, cause now beacause of changes there is realy big mess between branches ;)

cschindl commented:
Hi @wacki4 ,

I have now made the adjustments in the branch "feature/grouping-servers". Can you test everything again from your side?

wacki4 commented:
Only problem that i found - when i've added one folder, get to servers, set on some of them this folder, got back to folders and added child folder for previous one - it wasn't seen on folders in server config. after save and getting back to server config - it was seen correctly.

Also, when deleting folder with servers, and folder has parent - maybe better would be to move servers to parent folder? It should not be hard to done.
